src/api.proposed.x.d.ts

```diff
@@ -0,0 +1,171 @@
+         * I.e. the session names map to the relative path of the notebook file.
+         * As a result when attempting to create a new session for a notebook/file, Jupyter will
+         * first check if a session already exists for the same file and same kernel, and if so, will re-use that session.
+         */
+        readonly mappedRemoteNotebookDir?: Uri;
+        /**
+         * Returns the sub-protocols to be used. See details of `protocols` here https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket
+         * Useful if there is a custom authentication scheme that needs to be used for WebSocket connections.
+         * Note: The client side npm package @jupyterlab/services uses WebSockets to connect to remote Kernels.
+         */
+        readonly webSocketProtocols?: string[];
+    }
+
+    /**
+     * Represents a Jupyter Server displayed in the list of Servers.
+     */
+    export interface JupyterServer {
+        /**
+         * Unique identifier for this server.
+         */
+        readonly id: string;
+        /**
+         * A human-readable string representing the name of the Server. This can be read and updated by the extension.
+         */
+        label: string;
+    }
+
+    /**
+     * Provider of Jupyter Servers.
+     */
+    export interface JupyterServerProvider {
+        /**
+         * Event fired when the list of servers changes.
+         * Note: the getJupyterServers method will not be called unless changes are detected.
+         */
+        onDidChangeServers?: Event<void>;
+        /**
+         * Returns the list of servers.
+         */
+        getJupyterServers(token: CancellationToken): Promise<JupyterServer[]>;
+        /**
+         * Returns the connection information for the Jupyter server.
+         */
+        resolveConnectionInformation(
+            server: JupyterServer,
+            token: CancellationToken
+        ): Promise<JupyterServerConnectionInformation>;
+    }
+    /**
+     * Represents a reference to a Jupyter Server command.
+     */
+    export interface JupyterServerCommand {
+        /**
+         * Title of the command, like `save`.
+         */
+        title: string;
+        /**
+         * A human-readable string which is rendered less prominent in a separate line.
+         */
+        detail?: string;
+        /**
+         * A tooltip for the command, when represented in the UI.
+         */
+        tooltip?: string;
+        /**
+         * Default command to be used when there are no servers.
+         * If not set, and there are not servers, then the user will be prompted to select a command from a list of commands returned by `getCommands`.
+         */
+        picked?: boolean;
+    }
+    /**
+     * Provider of Jupyter Server Commands.
+     * Each command allows the user to perform an action.
+     */
+    export interface JupyterServerCommandProvider {
```

Comment:
I have trouble understanding why there is a need for a command provider and why it isn't just `JupyterServerCollection#commands: vscode.Command[]`. These commands can then do whatever is needed to establish a new server connection, be that nested quick picks or something else. 

Do you foresee these command to be changing a lot and be different each time the UI shows?

Reply:
No, I don’t see this changing very often,
It can change after users select a command, at least that’show code spaces seems to work


> believe a difference is that the remote repo picker can not list all remotes, that's why it reacts to what's being typed (so that it can query GH as you type). Do you envision a similar use-case and is that generic to all commands or just one?

Right now, I see this being specific for just one command, and it’s specific to Jupyter extension, hence happy to leave it as an internal api/implementation that won’t be used by others. 
Meaning i think leaving commands as a property, as suggested is totally fine

> JupyterServerCollection#commands: vscode.Command[].

one problem with ‘Command’ type is the fact that there’s no way to tell us which one is selected, we need a way for extensions to tell us which item is to be selected by default in some cases

the only use case is, if there’s only one command item and no servers and that command item has it’s ‘picked’ property set to ‘true’ then we automatically select that instead of displaying the quick pick to the user,

this is why I went with another type named JupyterServerCommand

The other reason for this type is to have a strong return type documented
When commands are handled 
we have a very specific return type, the return type when a command is handled and executed has three states, 
- user selected a server (hence server returned)   
- user wishes to go back to the previous ui (hit a back button in their extension ui)
- User hits esc key
- 

Spoke to @mjbvz and his suggestion was to go with tours approach which I like, but happy to change

Reply:
Hopefully the following is a better explanation of why we need this.
when a command is selected, then we need to call back into the 3rd party extension and the return value of the called function has the following signature

```typescript
        /**
         * Invoked when a command has been selected.
         * @param command The command selected by the user.
         * @returns
         * - JupyterServer  : The Jupyter Server object that was created
         * - 'back'         : Go back to the previous UI in the workflow
         * - undefined|void : Do nothing
         */
        handleCommand(
            command: JupyterServerCommand,
            token: CancellationToken
        ): Promise<JupyterServer | 'back' | undefined | void>;

```

Here's a sample UI displayed by the 3rd party extension when a command is selected.
As you can see there are 3 things users can do:
* Click back button (at that point, control should come back to the Jupyter extension and we need to re-display the same quick pick)
* User can hit `esc` key
* Or user can go through the custom UI and evntually that will yield a Jupyter Server.
![Screenshot 2023-08-18 at 14 40 14](https://github.com/microsoft/vscode-jupyter/assets/1948812/443cf3f9-0ce7-49c5-9baf-457656ba43d0)